src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    // to stop the completion session.
+                    if (_completionRules.DismissIfEmpty)
+                    {
+                        return null;
+                    }
+                }
+
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                var selection = UpdateSelectionHint.SoftSelected;
+
+                return new FilteredCompletionModel(
+                    ImmutableArray<CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                    _data.SelectedFilters, selection, centerSelection: true, uniqueItem: null);
+            }
+
+            private ImmutableArray<CompletionFilterWithState> GetUpdatedFilters(IReadOnlyList<MatchResult<VSCompletionItem>> items)
+            {
+                var showCompletionItemFilters = _globalOptions.GetOption(CompletionViewOptions.ShowCompletionItemFilters, _document?.Project.Language);
+                if (!showCompletionItemFilters)
+                    return ImmutableArray<CompletionFilterWithState>.Empty;
+
+                // See which filters might be enabled based on the typed code
+                using var _ = PooledHashSet<CompletionFilter>.GetInstance(out var textFilteredFilters);
+                textFilteredFilters.AddRange(items.SelectMany(n => n.EditorCompletionItem.Filters));
+
+                // When no items are available for a given filter, it becomes unavailable.
+                // Expanders always appear available as long as it's presented.
+                return _data.SelectedFilters.SelectAsArray(n => n.WithAvailability(n.Filter is CompletionExpander || textFilteredFilters.Contains(n.Filter)));
+            }
+
+            /// <summary>
+            /// Given multiple possible chosen completion items, pick the one that has the
+            /// best MRU index, or the one with highest MatchPriority if none in MRU.
+            /// </summary>
+            private static RoslynCompletionItem GetBestCompletionItemBasedOnMRU(
+                ImmutableArray<RoslynCompletionItem> chosenItems, ImmutableArray<string> recentItems)
+            {
+                Debug.Assert(chosenItems.Length > 0);
+
+                // Try to find the chosen item has been most recently used.
+                var bestItem = chosenItems[0];
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
+                    var chosenItem = chosenItems[i];
+                    var mruIndex1 = GetRecentItemIndex(recentItems, bestItem);
```

Comment:
this recomputes mruIndex every time through this loop.  we can just compute it when we set the best item (outside the loop, and in the if below).